apps/dotcom/sync-worker/src/TLPostgresReplicator.ts

```diff
@@ -61,36 +65,83 @@ export class TLPostgresReplicator extends DurableObject<Environment> {
 		})
 
 		this.ctx.blockConcurrencyWhile(async () => {
+			/**
+			 * BOOTUP SEQUENCE
+			 * 1. Generate a unique boot id
+			 * 2. Subscribe to all changes
+			 * 3. Fetch all data and write the boot id in a transaction
+			 * 4. If we receive events before the boot id update comes through, ignore them
+			 * 5. If we receive events after the boot id comes through but before we've finished
+			 *    fetching all data, buffer them and apply them after we've finished fetching all data.
+			 *    (not sure this is possible, but just in case)
+			 * 6. Once we've finished fetching all data, apply any buffered events
+			 */
 			let didFetchInitialData = false
+			let didGetBootIdInSubscription = false
+			const myId = this.ctx.id.toString()
+			const bootId = uniqueId()
 			// TODO: time how long this takes
 			await new Promise((resolve, reject) => {
+				const initialUpdateBuffer: Array<{
+					row: postgres.Row | null
+					info: postgres.ReplicationEvent
+				}> = []
 				this.db
 					.subscribe(
 						'*',
 						async (row, info) => {
-							if (!didFetchInitialData) {
-								return
-							}
-							try {
-								await this.handleEvent(row, info)
-							} catch (e) {
-								this.captureException(e)
+							if (didGetBootIdInSubscription) {
+								if (!didFetchInitialData) {
+									initialUpdateBuffer.push({ row, info })
+								} else {
+									try {
+										this.handleEvent(row, info)
+									} catch (e) {
+										this.captureException(e)
+									}
+								}
+							} else if (
+								info.relation.table === 'replicator_boot_id' &&
+								(row as any)?.bootId === bootId
+							) {
+								didGetBootIdInSubscription = true
+							} else {
+								// ignore events until we get the boot id
 							}
 						},
 						async () => {
 							try {
-								const users = await this.db`SELECT * FROM public.user`
-								const files = await this.db`SELECT * FROM file`
-								const fileStates = await this.db`SELECT * FROM file_state`
+								const sql = getFetchEverythingSql(myId, bootId)
 								this.ctx.storage.sql.exec(seed)
-								users.forEach((user) => this.insertRow(user, 'user'))
-								files.forEach((file) => this.insertRow(file, 'file'))
-								fileStates.forEach((fileState) => this.insertRow(fileState, 'file_state'))
+								// ok
+								await this.db.begin(async (db) => {
+									return db
+										.unsafe(sql, [], { prepare: false })
+										.simple()
+										.forEach((row: any) => {
+											this.insertRow(
```

Comment:
Might make sense to bulk insert here. We can do some tests when we'll test perf of other parts.

Reply:
yeah let's wait, sqlite writes are supposed to be basically instant on DOs